TT-7163 Add support for Resurface backend to pumps

monrax opened the pull request:
## Description
- Defines `ResurfacePump` inside `pumps/resurface.go`, and adds it to the stack in `pumps/init.go`
- The `ResurfacePump` populates http.Request and http.Response structs with (preferably raw) data obtained from each Tyk Analytics record, and uses the [logger-go](https://pkg.go.dev/github.com/resurfaceio/logger-go) module to send the data to Resurface.
- The `ResurfacePump` supports capture of custom Tyk fields, including API ID, Name, Version, Org ID, and Oauth ID.
- Test file and updated README are included.

## Related Issue
Issue #530
https://tyktech.atlassian.net/browse/TT-7163
## Motivation and Context
Extend the capabilities of the Tyk Gateway with an integration with [Resurface](https://resurface.io/) using the Tyk Pump.

## How This Has Been Tested
- Unit tests in `pumps/resurface_test.go`
- Local QA without compiling
- Local QA with compiling, using `Dockerfile.std` to build custom dev image
- Local QA using docker-compose spec included in [Tyk demo](https://github.com/TykTechnologies/tyk-pro-docker-demo)

## Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/7117255/203390413-31d51bd1-8907-4d55-960a-2e62433700ea.png)
![image](https://user-images.githubusercontent.com/7117255/203390799-23a18bf0-53b7-412e-85cf-d473217f54b2.png)


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply -->
<!-- If you're unsure about any of these, don't hesitate to ask; we're here to help! -->
- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [x] My change requires a change to the documentation.
  - [x] If you've changed APIs, describe what needs to be updated in the documentation.
- [x] I have updated the documentation accordingly.
- [x] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [x] `go vet`

tbuchaillot commented:
Hi @monrax ! Thanks a lot for the contribution! We will schedule some time to have a look at this and come back to you :)

tbuchaillot commented:
Hey @monrax! Do you guys have a demo or trial version of Resurface to validate this on our side?

monrax commented:
hey @tbuchaillot ! Apologies for the late reply.

You can [run Resurface using docker](https://resurface.io/docs#running-on-docker):

```bash
docker run -v resurface:/db -d --name resurface -p 7700:7700 -p 7701:7701 --cpus=4 --memory=8g resurfaceio/resurface:3.4.21
```

The UI can be accessed at http://localhost:7700. You should be greeted with a Log In screen (auth is disabled by default so you can use any username), then you'll be prompted to enter your license.

Alternatively, you can use [this modified `docker-compose.yml` file](https://gist.github.com/monrax/3e6042e716034315b55d9963bb74354b) that should allow you to run Resurface locally with the rest of the `tyk-pro-docker-demo`. The Tyk Pump portion is commented out so you can just clone our forked repo, checkout the `feature/resurface` branch, add the `pump.conf` file and do `go run .` to see it in action.

I took the liberty of sending an email to tombuchaillot89@gmail.com with the license. Please, let me know if this works for you.

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,217 @@
+
+	"github.com/TykTechnologies/tyk-pump/analytics"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
```
@monrax can you remove this dependency? Ideally you'd want to use Go errors pkg.

monrax replied:
Done

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,217 @@
+	}
+	logger, err := logger.NewHttpLogger(opt)
+	if err != nil {
+		log.Fatal(err)
```
@monrax why a `log.Fatal`? I think it would be better to use `rp.log.Error` and raise the error to the main execution process.

monrax replied:
You are right, apologies for the poor error handling. I've refactored it accordingly.

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,217 @@
+	}
+	parsedURL, err := url.Parse(schema + "://" + host + path)
+	if err != nil {
+		log.Fatal("Problem converting url", err)
```
@monrax can we remove the `log.Fatal` from here too? Why do we need to parse the URL?

monrax replied:
Thank you for pointing it out, I just refactored all error handling.

Internally, the Resurface backend uses our [logger-go](https://github.com/resurfaceio/logger-go) tool which takes care of parsing all available details from the HTTP transaction. In addition, it applies some [logging rules](https://resurface.io/docs#logging-rules), and submits the entire reconstructed message to the configured [Capture URL ](https://resurface.io/docs#getting-capture-url) to the running Resurface instance using a specific [JSON format](https://resurface.io/docs#json-format). The main use cases of the `logger-go` package are middleware implementations, so we chose both the `Request` and `Response` types from the `net/http` package as the vehicles to receive HTTP transaction data but, as long as the data can be added to those structs, `logger-go` can be used in other types of implementations (like a Tyk Pump backend!) just fine. All this is to say, the `URL` field in the `Request` struct is a [URL](https://pkg.go.dev/net/url#URL) type from the `net/url` package, so it has to be parsed in order to assign it to `httpReq.URL`.

tbuchaillot commented on pumps/resurface_test.go:
```diff
@@ -0,0 +1,333 @@
+	}
+
+	err := pmp.WriteData(context.TODO(), recs)
+	if err != nil {
```
We use `assert.NotNil` here

monrax replied:
Thank you, I just modified it accordingly.

tbuchaillot commented on pumps/resurface_test.go:
```diff
@@ -0,0 +1,333 @@
+
+	err := pmp.WriteData(context.TODO(), recs)
+	if err != nil {
+		t.Fatal(pmp.GetName()+"couldn't write records with err:", err)
```
we can use `assert.NotNil` here

monrax replied:
Thank you for pointing this out, I just modified it.

tbuchaillot commented on pumps/resurface_test.go:
```diff
@@ -0,0 +1,333 @@
+		},
+	})
+	if err != nil {
+		t.Fatal(pmp.GetName()+"couldn't write records with err:", err)
```
we can use assert.NotNil here

monrax replied:
Thank you, I just modified it.

tbuchaillot commented on pumps/resurface_test.go:
```diff
@@ -0,0 +1,333 @@
+
+	err := pmp.WriteData(context.TODO(), recs)
+	if err != nil {
+		t.Fatal(pmp.GetName()+"couldn't write records with err:", err)
```
we can use assert.NotNil here

monrax replied:
Thank you, I just modified it accordingly.

tbuchaillot commented on pumps/resurface_test.go:
```diff
@@ -0,0 +1,333 @@
+
+	err := pmp.Init(cfg)
+	if err != nil {
+		t.Fatal("Problem initializing "+pmp.GetName()+": ", err)
```
we can use assert.NotNil here

monrax replied:
Thank you, I just modified it accordingly.

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,217 @@
+
+	err := mapstructure.Decode(config, &rp.config)
+	if err != nil {
+		rp.log.Fatal("Failed to decode configuration: ", err)
```
I don't think we should `Fatal` here. Maybe just a debug log can be enough

monrax replied:
Thank you, you are absolutely right. I just modified this.

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,217 @@
+		"Org-ID",
+	}
+
+	// Decode raw HTTP transaction from base64 strings
```
@monrax what would happens if `rec.RawRequest` or `rec.RawResponse` are empty? Usually req/resp bodies are omitted in prod and just used for debugging https://tyk.io/docs/tyk-stack/tyk-pump/useful-debug-modes/#what-is-detailed-request-logging.

Wouldn't it be better to check the length of those fields first?

monrax replied:
As a data lake for API calls, Resurface is able to provide context-based security analysis for folks to make data-driven decisions. In order for this to happen, we need to capture the entire payload. Because of this, we assume all the raw data is passed to `mapRawData`. So, we do check for length for these fields, just in the `WriteData` function.

Having said that, is there a way to enable detailed logging just for Resurface? (like having an option similar to `omit_detailed_recording` set to `true` for all storage destinations by default, and only enabling a backend-specific option similar to `enable_detailed_recording` only for Resurface. Sort of the opposite of how detailed logging is configured in the `tyk.conf` file at the end of that documentation page you shared)

tbuchaillot replied:
> Having said that, is there a way to enable detailed logging just for Resurface?
@monrax I'm afraid that's not possible. Tyk gateway is not aware of what pumps are configured here, and it just moves the analytic data to Redis to be processed by Tyk Pump. 
You can enable detailed analytics per API or even per endpoint but not for X/Y data storage since it would imply that all the API analytic data is processed and stored in Redis for pump consumption.

tbuchaillot reviewed (changes requested):
I added some comments! Usually we try to avoid the usage of `log.Fatal` since it would end the execution of Tyk-Pump and we don't want that.

monrax commented on pumps/resurface.go:
```diff
@@ -0,0 +1,217 @@
+
+	// Request URL
+	schema := "http"               // TODO - how to determine schema from Tyk Analytics?
+	host := reqHeaders.Get("Host") // TODO - how to determine actual used hostname? (not Host header)
```
Thanks!

monrax commented:
Hola, @tbuchaillot. Happy New Year :D

Apologies for the late reply, the holiday break slowed everything down for a bit. We are back on the saddle, though!
I refactored the `mapRawData` function taking into account your comments and request for changes. We dropped `log.Fatal` entirely (sorry about it being everywhere, btw), and modified the logic so that it copies the available fields (for host, path, method, and response code) from the `AnalyticsRecord` first, and the raw HTTP request/response only when those are empty. I also updated the corresponding `resurface_test` and rebased it all. Please, let me know if I missed something or if there is anything else to modify/clarify.

tbuchaillot commented:
> Hola, @tbuchaillot. Happy New Year :D
> 
> Apologies for the late reply, the holiday break slowed everything down for a bit. We are back on the saddle, though! I refactored the `mapRawData` function taking into account your comments and request for changes. We dropped `log.Fatal` entirely (sorry about it being everywhere, btw), and modified the logic so that it copies the available fields (for host, path, method, and response code) from the `AnalyticsRecord` first, and the raw HTTP request/response only when those are empty. I also updated the corresponding `resurface_test` and rebased it all. Please, let me know if I missed something or if there is anything else to modify/clarify.

Thanks @monrax! I'll have a look and come back to you 👍

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,234 @@
+	for _, v := range data {
+		decoded := v.(analytics.AnalyticsRecord)
+		if len(decoded.RawRequest) == 0 && len(decoded.RawResponse) == 0 {
+			log.WithField("prefix", resurfacePrefix).Warn("Record dropped. Please enable Detailed Logging.")
```
this can be changed to `p.log.Warn`

tbuchaillot commented on pump.example.conf:
```diff
@@ -62,6 +62,13 @@
         "buffered_max_messages": 32
       }
+    },
+    "resurfaceio": {
```
@monrax can you remove the config from `pump.example.conf`?

tbuchaillot reviewed (changes requested):
Looks better ! Just some small changes, also please solve the conflicts 🙏

monrax commented:
> Looks better ! Just some small changes, also please solve the conflicts pray

Hey, @tbuchaillot. I just rebased these changes again on top of the latest upstream commits. In addition, I carried out the corresponding tests. As before, please, let me know if I missed something or if there is anything else to modify/clarify. Thanks!

tbuchaillot reviewed (approved):
LGTM 🚀

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,234 @@
+	p.log.Debug("Writing ", len(data), " records")
+
+	for _, v := range data {
+		decoded := v.(analytics.AnalyticsRecord)
```
It would be good to check the casting here. 
Something like this should do the trick:
```go
decoded, ok := v.(analytics.AnalyticsRecord)
if !ok {
   p.log.Error("error decoding analytic record")
   continue
}
```

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,234 @@
+	if path == "" {
+		path = rawPath
+	} else if idx := strings.Index(rawPath, "?"); idx != -1 {
+		path = path + rawPath[idx:]
```
this can be changed to `path += rawPath[idx:]`

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,234 @@
+	return
+}
+
+func mapRawData(rec analytics.AnalyticsRecord) (httpReq http.Request, httpResp http.Response, customFields map[string]string, err error) {
```
@monrax we should pass the `analytics.AnalyticsRecord` as a pointer here, it's a heavy struct.

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,234 @@
+	if rp.logger.Enabled() {
+		rp.log.Info(rp.GetName() + " Initialized")
+		return nil
+	} else {
```
if block ends with a return statement, so drop this else and outdent its block

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,234 @@
+
+type ResurfacePumpConfig struct {
+	EnvPrefix string `mapstructure:"meta_env_prefix"`
+	Url       string `mapstructure:"capture_url"`
```
lets change `Url` for `URL`

tbuchaillot commented on pumps/resurface.go:
```diff
@@ -0,0 +1,234 @@
+	return
+}
+
+func (p *ResurfacePump) WriteData(ctx context.Context, data []interface{}) error {
```
receiver name `p` should be consistent with previous receiver name `rp` for ResurfacePump

tbuchaillot commented on pumps/resurface_test.go:
```diff
@@ -0,0 +1,329 @@
+}
+
+func TestResurfaceWriteData(t *testing.T) {
+	const MOCK_HOST = "test0"
```
Can we use CamelCase here?

tbuchaillot reviewed (changes requested):
@monrax I ran the `golangci-lint` over this branch and some smalls things popped up 😅 Added the comments. 
Also the `resurface_test.go` file needs to be `gofumpt`'d with `-extra` flag (`gofumpt -w -l -extra pumps/resurface_test.go`)

Please feel free to run the linter locally with the following cmd:

`golangci-lint run --out-format checkstyle --timeout=300s --max-issues-per-linter=0 --max-same-issues=0 --issues-exit-code=0 ./...`

and re-request my review when everything is 👌

monrax commented:
hey @tbuchaillot I just pushed the modifications you requested. Thank you for those, btw!

I ran the linter and all the issues that show up are related to undeclared or undefined variables. These are not actually so, since they are declared/defined during initialization (or elsewhere in the `pumps` package):

```xml
<file name="pumps/resurface.go">
  <error column="2" line="22" message="undeclared name: `CommonPumpConfig`" severity="error" source="typecheck"></error>
  <error column="24" line="35" message="undeclared name: `PUMPS_ENV_PREFIX`" severity="error" source="typecheck"></error>
  <error column="32" line="38" message="undeclared name: `Pump`" severity="error" source="typecheck"></error>
  <error column="11" line="53" message="undeclared name: `log`" severity="error" source="typecheck"></error>
  <error column="6" line="57" message="rp.log undefined (type *ResurfacePump has no field or method log)" severity="error" source="typecheck"></error>
  <error column="2" line="61" message="undeclared name: `processPumpEnvVars`" severity="error" source="typecheck"></error>
  <error column="6" line="70" message="rp.log undefined (type *ResurfacePump has no field or method log)" severity="error" source="typecheck"></error>
  <error column="6" line="74" message="rp.log undefined (type *ResurfacePump has no field or method log)" severity="error" source="typecheck"></error>
  <error column="5" line="77" message="rp.log undefined (type *ResurfacePump has no field or method log)" severity="error" source="typecheck"></error>
  <error column="5" line="214" message="rp.log undefined (type *ResurfacePump has no field or method log)" severity="error" source="typecheck"></error>
  <error column="7" line="219" message="rp.log undefined (type *ResurfacePump has no field or method log)" severity="error" source="typecheck"> </error>
  <error column="7" line="223" message="rp.log undefined (type *ResurfacePump has no field or method log)" severity="error" source="typecheck"> </error>
  <error column="7" line="229" message="rp.log undefined (type *ResurfacePump has no field or method log)" severity="error" source="typecheck"> </error>
</file>
```

I also ran `gofumpt` on `resurface_tests.go` without any incidents. Apologies for all this back-and-forth thing. I just wanna make sure everything is up to standard. I believe everything is fine now. However, if I missed something, like always, please let me know! Thanks.

tbuchaillot reviewed (approved):
LGTM 🚀

tbuchaillot commented:
@monrax Merged 💪  Thanks a lot for the contribution! This will be included in April's next Tyk Pump release (v1.8).
